compiler/src/iree/compiler/Reducer/Strategies/ReduceFlowDispatchResultBySplat.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <random>
+
+#include "iree/compiler/Reducer/Strategies/DeltaStrategies.h"
+
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/Passes.h"
+
+using namespace mlir;
+using namespace mlir::iree_compiler;
+
+void mlir::iree_compiler::reduceFlowDispatchResultBySplatDelta(
+    ChunkManager &chunker, WorkItem &workItem) {
+  ModuleOp module = workItem.getModule();
+
+  // Create a list of dispatch ops we want to replace.
+  SmallVector<IREE::Flow::DispatchOp> dispatchOps;
+  SmallVector<IREE::Flow::DispatchOp> keepOps;
+  module.walk([&](IREE::Flow::DispatchOp dispatchOp) {
+    bool shouldReplace = true;
+    for (auto result : dispatchOp.getResults()) {
+      // Only replace tensor types.
+      if (!isa<RankedTensorType>(result.getType())) {
+        shouldReplace = false;
+        break;
+      }
+    }
+
+    if (shouldReplace && !chunker.shouldFeatureBeKept()) {
+      dispatchOps.push_back(dispatchOp);
+    } else {
+      keepOps.push_back(dispatchOp);
+    }
+  });
+
+  if (dispatchOps.empty()) {
+    return;
+  }
+
+  OpBuilder builder(module.getContext());
+
+  // Insert util.optimization_barrier for results of keep dispatch ops.
+  for (auto dispatchOp : keepOps) {
+    builder.setInsertionPointAfter(dispatchOp);
+    for (Value result : dispatchOp.getResults()) {
```

Comment:
So this is always all the results? (Trying to think if that is too conservative)

Reply:
Yes. I plan to add a delta pass which will start removing optimization barriers to check if we still can produce the test behavior we need.